taroscript/tx.go

```diff
@@ -327,35 +329,123 @@ func InputScriptSpendSigHash(virtualTx *wire.MsgTx, input *asset.Asset,
-	inputAsset *asset.Asset, idx int, sigHashType txscript.SigHashType,
-	txSigner Signer) (*wire.TxWitness, error) {
+// CreateTaprootSignature creates a Taproot signature for the given asset input.
+// Depending on the fields set in the input, this will either create a key path
+// spend or a script path spend.
+func CreateTaprootSignature(vIn *taropsbt.VInput, virtualTx *wire.MsgTx,
+	idx int, txSigner Signer) (wire.TxWitness, error) {
+
+	// Before we even attempt to sign anything, we need to make sure all the
+	// input information we require is present.
+	if len(vIn.TaprootBip32Derivation) == 0 {
+		return nil, fmt.Errorf("missing input Taproot BIP32 derivation")
+	}
+
+	// Currently, we only support creating one signature per input.
+	//
+	// TODO(guggero): Should we support signing multiple paths at the same
+	// time? What are the performance and security implications?
+	if len(vIn.TaprootBip32Derivation) > 1 {
+		return nil, fmt.Errorf("unsupported multiple taproot BIP32 " +
+			"derivation info found, can only sign for one at a" +
+			"time")
+	}
+	if len(vIn.TaprootBip32Derivation[0].LeafHashes) > 1 {
+		return nil, fmt.Errorf("unsupported number of leaf hashes in " +
+			"taproot BIP32 derivation info, can only sign for one " +
+			"at a time")
+	}
+
+	derivation := vIn.TaprootBip32Derivation[0]
 
 	// Compute a virtual prevOut from the input asset for the signer.
-	prevOut, err := InputAssetPrevOut(*inputAsset)
+	prevOut, err := InputAssetPrevOut(*vIn.Asset())
 	if err != nil {
 		return nil, err
 	}
 
-	// Build the signing descriptor for a BIP 86 signature over the Taro
-	// virtual TX. To follow BIP 86, no tweaks nor witness script are
-	// provided, and the untweaked internal key is used as the signing key.
+	// Start with a default sign descriptor and the BIP 86 sign method then
+	// adjust depending on the input parameters.
 	spendDesc := lndclient.SignDescriptor{
 		KeyDesc: keychain.KeyDescriptor{
-			PubKey: &internalKey,
+			PubKey: vIn.Asset().ScriptKey.RawKey.PubKey,
 		},
 		SignMethod: input.TaprootKeySpendBIP0086SignMethod,
 		Output:     prevOut,
-		HashType:   sigHashType,
+		HashType:   vIn.SighashType,
 		InputIndex: idx,
 	}
 
+	// There are three possible signing cases: BIP 86 key spend path, key
+	// spend path with a script root, and script spend path.
+	switch {
+	// If there is no merkle root, we're doing a BIP 86 key spend.
+	case len(vIn.TaprootMerkleRoot) == 0:
+		// This is the default case, so we don't need to do anything.
+
+	// No leaf hash means we're not signing a specific script, so this is
+	// the key spend path with a script root.
+	case len(vIn.TaprootMerkleRoot) == sha256.Size &&
+		len(derivation.LeafHashes) == 0:
+
+		spendDesc.SignMethod = input.TaprootKeySpendSignMethod
+		spendDesc.TapTweak = vIn.TaprootMerkleRoot
+
+	// One leaf hash and a merkle root means we're signing a specific
+	// script.
+	case len(vIn.TaprootMerkleRoot) == sha256.Size &&
+		len(derivation.LeafHashes) == 1:
```

Comment:
Does this restrict to only a control block of size 1? So you only have a single sibling.

Reply:
No, we just only support signing for a single leaf at a time. The PSBT spec allows you to specify multiple leaf hashes and that would indicate to the signer that it should create a signature for each of them. But since we always know which path we're going to use, we haven't implemented this in `lnd` either. I updated the comment to make it more clear.